packages/compile-common/src/profiler/updated.ts

```diff
@@ -0,0 +1,157 @@
+  });
+  // Get all the artifact files, and read them, parsing them as JSON
+  let buildFiles: string[];
+  try {
+    buildFiles = fse.readdirSync(contracts_build_directory);
+  } catch (error) {
+    // The build directory may not always exist.
+    if (error.message.includes("ENOENT: no such file or directory")) {
+      // Ignore it.
+      buildFiles = [];
+    } else {
+      throw error;
+    }
+  }
+
+  buildFiles = buildFiles.filter(file => path.extname(file) === ".json");
+  const jsonData = buildFiles.map(file => {
+    const body = fse.readFileSync(
+      path.join(contracts_build_directory, file),
+      "utf8"
+    );
+    return { file, body };
+  });
+
+  for (let i = 0; i < jsonData.length; i++) {
+    try {
+      const data: ContractObject = JSON.parse(jsonData[i].body);
+
+      // In case there are artifacts from other source locations.
+      if (sourceFilesArtifacts[data.sourcePath] == null) {
+        sourceFilesArtifacts[data.sourcePath] = [];
+      }
+
+      sourceFilesArtifacts[data.sourcePath].push(data);
+    } catch (error) {
+      // JSON.parse throws SyntaxError objects
+      if (error instanceof SyntaxError) {
+        throw new Error("Problem parsing artifact: " + jsonData[i].file);
+      } else {
+        throw error;
+      }
+    }
+  }
+  return sourceFilesArtifacts;
+}
+
+function findUpdatedFiles(
+  sourceFilesArtifacts: SourceFilesArtifacts,
+  sourceFilesArtifactsUpdatedTimes: SourceFilesArtifactsUpdatedTimes
+): string[] {
+  // Stat all the source files, getting there updated times, and comparing them to
+  // the artifact updated times.
+  const sourceFiles = Object.keys(sourceFilesArtifacts);
+
+  let sourceFileStats: (fse.Stats | null)[];
+  sourceFileStats = sourceFiles.map(file => {
+    try {
+      return fse.statSync(file);
+    } catch (error) {
+      // Ignore it. This means the source file was removed
+      // but the artifact file possibly exists. Return null
+      // to signfy that we should ignore it.
+      return null;
+    }
+  });
+
+  return sourceFiles
+    .map((sourceFile, index) => {
+      const sourceFileStat = sourceFileStats[index];
+
+      // Ignore updating artifacts if source file has been removed.
+      if (sourceFileStat == null) return;
+
+      const artifactsUpdatedTime =
+        sourceFilesArtifactsUpdatedTimes[sourceFile] || 0;
+      const sourceFileUpdatedTime = (
+        sourceFileStat.mtime || sourceFileStat.ctime
+      ).getTime();
+
+      if (sourceFileUpdatedTime > artifactsUpdatedTime) return sourceFile;
+    })
+    .filter(file => file);
+}
+
+function minimumUpdatedTimePerSource(
+  sourceFilesArtifacts: SourceFilesArtifacts
+) {
+  let sourceFilesArtifactsUpdatedTimes: SourceFilesArtifactsUpdatedTimes = {};
+  // Get the minimum updated time for all of a source file's artifacts
+  // (note: one source file might have multiple artifacts).
+  Object.keys(sourceFilesArtifacts).forEach(sourceFile => {
```

Comment:
Will have to run some recent perf tests but I think Object.keys with a plain old for loop is actually 20% faster on average than a forEach.